src/Utilities/Compiler/Options/SymbolNamesOption.cs

```diff
@@ -98,32 +154,145 @@ private SymbolNamesOption()
             {
                 return Empty;
             }
 
-            return new SymbolNamesOption(namesBuilder.ToImmutableDictionaryAndFree(), symbolsBuilder.ToImmutableDictionaryAndFree());
+            return new SymbolNamesOption(namesBuilder.ToImmutableDictionaryAndFree(), symbolsBuilder.ToImmutableDictionaryAndFree(),
+                wildcardNamesBuilder.ToImmutableDictionaryAndFree(x => x.Key, x => x.Value.ToImmutableDictionaryAndFree(), wildcardNamesBuilder.Comparer));
         }
 
         public bool IsEmpty => ReferenceEquals(this, Empty);
 
         public bool Contains(ISymbol symbol)
-            => _symbols.ContainsKey(symbol) || _names.ContainsKey(symbol.Name);
+            => _symbols.ContainsKey(symbol) || _names.ContainsKey(symbol.Name) || TryGetFirstWildcardMatch(symbol, out _);
 
         /// <summary>
         /// Gets the value associated with the specified symbol in the option specification.
         /// </summary>
-        public bool TryGetValue(ISymbol symbol, [NotNullWhen(true)] out string? value) =>
-            _symbols.TryGetValue(symbol, out value) || _names.TryGetValue(symbol.Name, out value);
+        public bool TryGetValue(ISymbol symbol, [NotNullWhen(true)] out string? value)
+        {
+            if (_symbols.TryGetValue(symbol, out value) || _names.TryGetValue(symbol.Name, out value))
+            {
+                return true;
+            }
+
+            if (TryGetFirstWildcardMatch(symbol, out var match))
+            {
+                value = match.Value;
+                return true;
+            }
+
+            value = null;
+            return false;
+        }
 
         public override bool Equals(object obj) => Equals(obj as SymbolNamesOption);
 
         public bool Equals(SymbolNamesOption? other)
-            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols);
+            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols) && _wildcardNamesBySymbolKind.IsEqualTo(other._wildcardNamesBySymbolKind);
 
         public override int GetHashCode()
-            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols));
+            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols), HashUtilities.Combine(_wildcardNamesBySymbolKind));
+
+        private bool TryGetFirstWildcardMatch(ISymbol symbol, out KeyValuePair<string, string?> firstMatch)
+        {
+            firstMatch = NoWildcardMatch;
+
+            if (_wildcardNamesBySymbolKind.IsEmpty)
+            {
+                return false;
+            }
+
+            Debug.Assert(symbol.Kind == SymbolKind.Event || symbol.Kind == SymbolKind.Field || symbol.Kind == SymbolKind.Method || symbol.Kind == SymbolKind.NamedType || symbol.Kind == SymbolKind.Namespace || symbol.Kind == SymbolKind.Property);
```

Comment:
I rely on the symbol kind but wasn't sure if it was best to play defensive and go for a bail-out condition.